### PR TITLE
Add `Reverse` type

### PR DESCRIPTION
**Pull request by benzaria**
- [`Reverse`](source/reverse.d.ts) - Creates a new array type by Reversing the order of each element in the original array.

### CODE REVIEW

File: source/reverse.d.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	Whether to preserve the optional modifier (`?`).
+
+	- When set to `true` and all members are optional, the optional modifiers are preserved as-is. but if one of the members is required the (`?`) is replaced with `| undefined`.
+	- When set to `false`, optional elements like `T?` are transformed to `T | undefined` or simply `T` depending on the `exactOptionalPropertyTypes` compiler option.
+
+	@example
+	```
+	import type {Reverse} from 'type-fest';
+
+	type Ex1 = Reverse<[string, number, boolean]>;
```
Comment (sindresorhus) [inline source/reverse.d.ts]: Don't use abbreviations.

File: source/reverse.d.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	*/
+	preserveOptionalModifier?: boolean;
+};
+
+type DefaultReverseOptions = {
+	preserveOptionalModifier: false;
+};
+
+/**
+Reverse an Array Items.
```
Comment (sindresorhus) [inline source/reverse.d.ts]:
```suggestion
Reverse array/tuple elements.
```

File: source/reverse.d.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+					? KeepOptional extends false
+						? [If<IsExactOptionalPropertyTypesEnabled, First, First | undefined>] // Add `| undefined` for optional elements, if `exactOptionalPropertyTypes` is disabled.
+						: [First?]
+					: [First],
+				...Tail,
+			]>
+			: never;
+
+/**
+Creates a new array type by Reversing the order of each element in the original array.
```
Comment (sindresorhus) [inline source/reverse.d.ts]:
```suggestion
Creates a new array type by reversing the element order of the original array.
```

File: source/reverse.d.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+/**
+Reverse options.
+
+@see {@link Reverse}
+*/
+type ReverseOptions = {
+	/**
+	Whether to preserve the optional modifier (`?`).
+
+	- When set to `true` and all members are optional, the optional modifiers are preserved as-is. but if one of the members is required the (`?`) is replaced with `| undefined`.
```
Comment (sindresorhus) [inline source/reverse.d.ts]:
```suggestion
	- When set to `true` and all members are optional, the optional modifiers are preserved as-is. but if any member is required the (`?`) is replaced with `| undefined`.
```

**Comment by sindresorhus** [comment]
- Export is missing. Add Reverse to index.d.ts
- Needs to be added to readme
- Should it be called `Reverse` or `ReverseTuple`?

**Comment by benzaria** [comment]
Hey @sindresorhus, it's been a while.
I will address the suggested changes and I think the name `Reverse` is fitting cuz its for arrays and tuples so using `ReverseTuple` will be misleading

File: source/reverse.d.ts (hunk 1 above)
Comment (benzaria) [reply source/reverse.d.ts]: sure thing

File: source/reverse.d.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+/**
+Reverse array/tuple elements.
+*/
+type _Reverse<
+	Array_ extends UnknownArray,
+	KeepOptional extends boolean,
+	Head extends UnknownArray = [],
+	Tail extends UnknownArray = [],
+> =
+	keyof Array_ & `${number}` extends never // Is `Array_` leading a rest element or empty
```
Comment (sindresorhus) [inline source/reverse.d.ts:69]: Would be useful with a short code comment that describes the algorithm.

File: test-d/reverse.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+declare const never: never;
+declare const string: string;
+declare const number: number;
+declare const symbol: symbol;
+declare const boolean: boolean;
+
+// Edge cases
+expectType<Reverse<[]>>([]);
+expectType<Reverse<any>>(any);
+expectType<Reverse<never>>(never);
```
Comment (sindresorhus) [inline test-d/reverse.ts]:
```suggestion
expectType<Reverse<never>>(never);
expectType<Reverse<[unknown]>>([{} as unknown] as const);
```

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/type-fest/pull/1166#issuecomment-2972873246 was merged. You mentioned you needed it for this.

File: test-d/reverse.ts [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+expectType<Reverse<[1, boolean, 'x']>>(['x', boolean, 1] as const);
+expectType<Reverse<[number, string, boolean]>>([boolean, string, number] as const);
+expectType<Reverse<[1, 2, 3, 4]>>([4, 3, 2, 1] as const);
+expectType<Reverse<[0, 1, 2, 3, 4]>>([4, 3, 2, 1, 0] as const);
+expectType<Reverse<['x', 1, true]>>([true, 1, 'x'] as const);
+
+// Optional/undefined
+expectType<Reverse<[1?, 2?, 3?]>>([3, 2, 1] as const);
+expectType<Reverse<[1?, 2?, 3?], {preserveOptionalModifier: true}>>({} as [3?, 2?, 1?]);
+expectType<Reverse<[1, 2?, 3?], {preserveOptionalModifier: true}>>({} as [3 | undefined, 2 | undefined, 1]);
```
Comment (som-sm) [inline test-d/reverse.ts]: Behaviour with optional elements is _not_ correct, you can't return `[3 | undefined, 2 | undefined, 1]` for `Reverse<[1, 2?, 3?]>` because at runtime input could be `[1, 2]` and reversing it would produce `[2, 1]` which wouldn't conform to `[3 | undefined, 2 | undefined, 1]`. 

So, the output for cases containing optional elements needs fixing.

**Comment by benzaria** [comment]
> @som-sm
> Yeah, please go ahead, we can close this PR and move ahead with yours if that feels better.

I’ve managed to:

* Incorporate the suggested change by adding `TailAcc |` to the `_Reverse` recursive call.
* Fix the case where all elements are optional.
* Remove `ReverseOptions` and refactor the implementation for clarity.
* Update the JSDoc and tests to reflect the new behavior.

If there are any missing tests or additional scenarios worth covering, feel free to list or add them.

**Comment by som-sm** [comment]
> Fix the case where all elements are optional.

@benzaria Please check https://github.com/sindresorhus/type-fest/pull/1266#issuecomment-3393452518.

**Comment by benzaria** [comment]
> @benzaria Please check [#1266 (comment)](https://github.com/sindresorhus/type-fest/pull/1266#issuecomment-3393452518).

@som-sm I took a note of that and I fixed the problem.

File: readme.md [hunk 8]
```diff
@@ -248,6 +248,7 @@ Click the type names for complete docs.
 - [`Arrayable`](source/arrayable.d.ts) - Create a type that represents either the value or an array of the value.
 - [`Includes`](source/includes.d.ts) - Returns a boolean for whether the given array includes the given item.
 - [`Join`](source/join.d.ts) - Join an array of strings and/or numbers using the given string as a delimiter.
+- [`Reverse`](source/reverse.d.ts) - Reverses the order of elements in a tuple or array type.
```
Comment (som-sm) [inline readme.md]: This should match the JSDoc description.

File: source/reverse.d.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+type T3 = Reverse<[string, number?, ...boolean[]]>;
+//=> [...boolean[], number, string] | [string]
+
+type T4 = Reverse<[string?, number?, ...boolean[]]>;
+//=> [...boolean[], number, string] | [string] | []
+```
+
+@category Array
+*/
+export type Reverse<Array_ extends UnknownArray> =
```
Comment (som-sm) [inline source/reverse.d.ts]: Better to name the type `ArrayReverse` as mentioned in #1266.

> I've kept the name as `ArrayReverse` instead of just `Reverse` to match with our existing `Array*` types (like `ArraySlice`, `ArraySplice`).

File: test-d/reverse.ts
Comment (som-sm) [inline test-d/reverse.ts:1]: Cases like labelled tuples, non-tuple arrays seem missing. Can you once cross-check from the tests in #1266.

File: readme.md (hunk 8 above)
Comment (benzaria) [reply readme.md]: they are the same, both in JSDoc and readme!

File: source/array-reverse.d.ts [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+			: [...HeadAcc, ...Array_, ...TailAcc] // Add the rest element between the accumulated elements.
+		: Array_ extends readonly [(infer First)?, ...infer Rest]
+			// Accumulate elements before a rest element in Arrayreverse order.
+			? IsOptionalKeyOf<Array_, 0> extends true
+				? (
+					| TailAcc // Union all possible cases when optional elements exist.
+					| _ArrayReverse<Rest, HeadAcc, [
+						If<IsExactOptionalPropertyTypesEnabled, First, First | undefined>, // Add `| undefined` for optional elements, if `exactOptionalPropertyTypes` is disabled.
+						...TailAcc,
+					]>
```
Comment (som-sm) [inline source/array-reverse.d.ts:81]: This breaks tail recursion, the existing long tuple test cases don't catch this because they _only_ contain required elements. If you instantiate with `Partial<TupleOf<50, number>>`, you'll see that the type breaks with a recursion depth error.

Refer #1266 for a tail recursive implementation. You can use the following test case for verification: 
```ts
expectType<TupleOf<IntRange<0, 51>, number>>({} as ArrayReverse<Partial<TupleOf<50, number>>>);
```
